File: snippets/java/src/main/java/develop/ReviewWorkflow.java
package develop;

import static develop.utils.Utils.askReview;
import static develop.utils.Utils.processReview;

import dev.restate.sdk.Restate;
import dev.restate.sdk.annotation.Shared;
import dev.restate.sdk.annotation.Workflow;
import dev.restate.sdk.common.DurablePromiseKey;

// <start_here>
@Workflow
public class ReviewWorkflow {
  // <start_promise_key>
  private static final DurablePromiseKey<String> REVIEW_PROMISE =
      DurablePromiseKey.of("review", String.class);

  // <end_promise_key>

  @Workflow
  public String run(String documentId) {
    // Send document for review
    Restate.run("ask-review", () -> askReview(documentId));

    // Wait for external review submission
    // <start_promise>
    String review = Restate.promise(REVIEW_PROMISE).future().await();
    // <end_promise>

    // Process the review result
    return processReview(documentId, review);
  }

  @Shared
  public void submitReview(String review) {
    // Resolve the workflow promise awaited by the run handler
    // <start_resolve_promise>
    Restate.promiseHandle(REVIEW_PROMISE).resolve(review);
    // <end_resolve_promise>
  }
}
// <end_here>

File: snippets/kotlin/src/main/kotlin/develop/ReviewWorkflow.kt
package develop

import dev.restate.sdk.annotation.Shared
import dev.restate.sdk.annotation.Workflow
import dev.restate.sdk.kotlin.*

// <start_here>
@Workflow
class ReviewWorkflow {

  companion object {
    // <start_promise_key>
    val REVIEW_PROMISE = durablePromiseKey<String>("review")
    // <end_promise_key>
  }

  @Workflow
  suspend fun run(documentId: String): String {
    // Send document for review
    runBlock { askReview(documentId) }

    // Wait for external review submission
    // <start_promise>
    val review: String = promise(REVIEW_PROMISE).future().await()
    // <end_promise>

    // Process the review result
    return processReview(documentId, review)
  }

  @Shared
  suspend fun submitReview(review: String) {
    // Resolve the workflow promise awaited by the run handler
    // <start_resolve_promise>
    promiseHandle(REVIEW_PROMISE).resolve(review)
    // <end_resolve_promise>
  }
}

// <end_here>

public suspend fun askReview(documentId: String): String {
  // Simulate sending document for review
  // In a real scenario, this could be an email, message queue, etc.
  return "Review requested for document $documentId"
}

public fun processReview(documentId: String, review: String): String {
  // Process the review result
  return "Processed review for document $documentId: $review"
}
